pkg/controller/cache/claim.go

```diff
@@ -36,15 +37,11 @@ import (
 	gcp "github.com/crossplaneio/stack-gcp/pkg/clients"
 )
 
-// A CloudMemorystoreInstanceClaimSchedulingController reconciles RedisCluster
-// claims that include a class selector but omit their class and resource
-// references by picking a random matching CloudMemorystoreInstanceClass, if
-// any.
-type CloudMemorystoreInstanceClaimSchedulingController struct{}
-
-// SetupWithManager sets up the
-// CloudMemorystoreInstanceClaimSchedulingController using the supplied manager.
-func (c *CloudMemorystoreInstanceClaimSchedulingController) SetupWithManager(mgr ctrl.Manager) error {
+// SetupCloudMemorystoreInstanceClaimSchedulingController adds a controller
+// that reconciles RedisCluster claims that include a class selector but omit
+// their class and resource references by picking a random matching
+// CloudMemorystoreInstanceClass, if any.
+func SetupCloudMemorystoreInstanceClaimSchedulingController(mgr ctrl.Manager, _ logging.Logger) error {
```

Comment:
Interesting. So **every** method will just take a logger now? Even if it isn't using it? How did we decide to add a `logging.Logger` parameter to these methods, but not to the ones later in the file, such as `ConfigureCloudMemorystoreInstance`?

Reply:
No, I don't intend for every method to take a logger.

To recap, the services end of Crossplane is architected as a series of common reconciler implementations with pluggable logic. The pluggable parts are typically types that satisfy a fairly specific interface in order to do a specific task (e.g. bind A to B, add a finalizer, or publish connection details). The reconcilers (e.g. `managed.Reconciler`, `claimbinding.Reconciler`) are defined in crossplane-runtime. _Most_ of the pluggable interfaces are defined in crossplane-runtime - the `Configurator` and `ExternalClient` implementations are the main exceptions to this rule.

My thinking is:

1. Each Crossplane binary should (only) instantiate the logger in `main.go`.
1. Loggers should be plumbed down to each reconciler.
1. Loggers may _sparingly_ be plumbed down to the pluggable interfaces.

Regarding point three - _most_ pluggable interfaces are simple, tightly scoped methods that are only 10-20 lines long. For such methods I think it makes sense for the reconciler invoking them to handle their logging, treating them almost like a third party library. If the method could fail in various ways it makes sense to add error context inside the method, allowing the logging to happen at the reconciler. If we're tempted to add "I am about to do the foo", "I successfully did the foo" style debug logs to such methods we may want to consider whether the methods can be factored down to simpler methods in order to keep the logging at the reconciler level. I think complex interfaces like `ExternalClient` will be the exception to this rule.

These `SetupTheFooController` functions are pure scaffolding. In all cases they:

1. Instantiate any non-default pluggable interfaces.
1. Instantiate the reconciler, passing in any interfaces from step 1.
1. Instantiate the controller, passing in the reconciler from step 2.
1. Add it to the manager.

Therefore I think it makes sense to plumb the logger down to every `SetupFooController` implementation as a required argument. The only reason the logger is not plumbed down in most cases right now is because this PR is still in draft. I would expect to plumb the logger down to every reconciler before taking this PR out of draft.